torch/nn/modules/_functions.py

```diff
@@ -0,0 +1,89 @@
+import torch
+from torch.autograd.function import Function
+
+
+class SyncBatchNorm(Function):
+
+    @staticmethod
+    def forward(self, input, weight, bias, running_mean, running_var, eps, momentum, process_group, world_size):
+        input = input.contiguous()
+
+        # calcualte mean/invstd for input.
+        mean, invstd = torch.batch_norm_stats(input, eps)
+
+        mean_all = torch.empty(world_size, mean.size(0), dtype=mean.dtype, device=mean.device)
+        invstd_all = torch.empty(world_size, invstd.size(0), dtype=invstd.dtype, device=invstd.device)
+        mean_l = list(mean_all.unbind(0))
+        invstd_l = list(invstd_all.unbind(0))
+        # using all_gather instead of all reduce so we can calculate mean/var in one go
+        torch.distributed.all_gather(mean_l, mean, process_group)
+        torch.distributed.all_gather(invstd_l, invstd, process_group)
+
+        # calcualte global mean & invstd
+        mean, invstd = torch.batch_norm_gather_stats(
+            input,
+            mean_all,
+            invstd_all,
+            running_mean,
+            running_var,
+            momentum,
+            eps,
+            int(input.numel() / input.size(1))
+        )
+
+        self.save_for_backward(input, weight, mean, invstd)
+        self.process_group = process_group
+        self.world_size = world_size
+
+        # apply element-wise normalization
+        out = torch.batch_norm_elemt(input, weight, bias, mean, invstd, eps)
+        return out
+
+    @staticmethod
+    def backward(self, grad_output):
+        grad_output = grad_output.contiguous()
+        saved_input, weight, mean, invstd = self.saved_tensors
+        grad_input = grad_weight = grad_bias = None
+        process_group = self.process_group
+        world_size = self.world_size
+
+        # calculate local stats as well as grad_weight / grad_bias
+        mean_dy, mean_dy_xmu, grad_weight, grad_bias = torch.batch_norm_backward_reduce(
+            grad_output,
+            saved_input,
+            mean,
+            invstd,
+            self.needs_input_grad[0],
+            self.needs_input_grad[1],
+            self.needs_input_grad[2]
+        )
+
+        if self.needs_input_grad[0]:
+            # synchronizing stats used to calculate input gradient.
+            # TODO: move div_ into batch_norm_backward_elemt kernel
+            torch.distributed.all_reduce(
+                mean_dy, torch.distributed.ReduceOp.SUM, process_group)
+            mean_dy.div_(world_size)
+            torch.distributed.all_reduce(
+                mean_dy_xmu, torch.distributed.ReduceOp.SUM, process_group)
+            mean_dy_xmu.div_(world_size)
```

Comment:
These allreduce calls can interfere with ones kicked off by DDP itself.

If autograd runs with single threaded with deterministic ordering you'll be fine, but as soon as it doesn't (e.g. there are multiple branches of the forward graph where the backward functions can be called in parallel), you'll run into deadlocks. This can be avoided by creating a new process group from the main one with `new_group` and using that throughout. Note that having multiple of these sync batch norm layers running `backward` in parallel can still deadlock, or worse, result in mixed up data, so for guaranteed correctness you'll have to use a separate process group per sync batch norm layer. This is not ideal and we may need to find a different solution for this.

cc @mrshenli @teng-li

Reply:
This is a gotcha for me. Thanks a lot for pointing out the issue. I don't fully understand how the allreduce call would cause trouble in branching while DDP handles that fine. Maybe I'll ask more details about this in private channel.

Just to reiterate, for the time being, it's a safe WAR as long as I have a separate process group per sync batch norm layer and use it for *both* forward and backward pass.
I'll copy/create a process group inside the initializer of SyncBatchNorm.

Reply:
Realized that duplicating process_groups in the initializer of SyncBatchNorm would not work.

As new_group should be called by all processes in the main group, and inside the initializer or converter function, each process would only see the given process_group it belongs to for that layer.

Reply:
I have this exact problem now!
My model has many SyncBNs in parallel (I also have an atypical module where I call autograd.grad() repeatedly on two tensors whose graph has syncBNs). In short the `backward()` has many `all_reduce` calls and some run in parallel. With DDP the code hangs without error and with NCCL logs I can see that it is because `all_reduce` hangs. 

I have tried torch native SyncBN, apex syncBN, and @ppwwyyxx 's NaiveSyncBN. This happens in both torch DDP and apex DDP (even with `delay_allreduce=True`). 

Is there something I can do to fix it? @jjsjann123 @pietern